### ISSUE TITLE
Fix Cassandra example umbrella issue

### ISSUE THREAD

**Issue by bgrant0607** [kubernetes/kubernetes#20961]
examples/cassandra is broken:

#20999
#19098
#19071
#17728
#17275
#10931

There text explaining the single-replica RC is also confusing: #20746

Also, the example currently reads endpoints. It could just use DNS.
https://github.com/kubernetes/kubernetes/blob/master/examples/cassandra/java/src/io/k8s/cassandra/KubernetesSeedProvider.java

We should check other best practices, such as putting multiple resources into a single file: #11361 

And should be updated to use Deployment, and possibly PetSet once that is available.

We should consider use of persistent volumes.

We should decide whether we want to show adoption in our main examples or not. #15193

The TL;DR section could go at the beginning, as in guestbook, and could collapse all the steps into 1.

The example doesn't really show usage of cassandra: #8355

**Comment by bgrant0607** [comment]
Note that cassandra e2e has been failing for weeks: #20999

**Comment by bgrant0607** [comment]
PR #20746 will revert the change to image v7.

**Comment by bgrant0607** [comment]
We're going to try to get this in shape to highlight on the website.

**Comment by bgrant0607** [comment]
Thread that suggests the example may not be running cassandra as recommended: https://groups.google.com/forum/#!topic/google-containers/87xj2CFMd2Y

**Comment by bgrant0607** [comment]
http://stackoverflow.com/questions/21776152/how-seed-node-works-in-cassandra-cluster

**Comment by bgrant0607** [comment]
Maybe it's ok, but perhaps we need to explain why in the example.

**Comment by amygdala** [comment]
Just a note on this umbrella issue that I'm currently working on "phase 2" of the cassandra example overhaul.  Will have another PR soon.  

(Perhaps we should fix the seed node usage issue as part of a presumed further upcoming overhaul to use PetSets).

**Comment by bgrant0607** [comment]
@chrislovecnm pointed out on slack:

not seeing instructions on how to recover a node within a kubernetes cluster.  You have to typically tweak cassandra.yaml.  http://docs.datastax.com/en/cassandra/2.0/cassandra/operations/ops_replace_node_t.html

**Comment by chrislovecnm** [comment]
Let me know how I can help out, don't have much time, but would like to assist!  Love kubernetes and really appreciate the work you folks are doing.

BTW you probably know but C* is shorthand for Cassandra.

**Node Recovery / Replacement**

I am coming up with three different use cases for recovery:

1. C* is not using persistent disk, and we need to bootstrap.  What are the instructions to do this following the procedure listed here: http://docs.datastax.com/en/cassandra/2.0/cassandra/operations/ops_replace_node_t.html
2. C* is using persistent disk and we need to recover a node using the disk that exists.  I am checking with some peeps at Datastax on how to do this.
3. C* using persistent disk, but the files have corrupted on the recovered disk, so we don't want to use that disk.  Pretty much like the first use case, but maybe we should throw it in the docs.

If you lose a seed node the recovery process is different as well.

**Persistent Disk**

That brings up the use case of using persistent disk with C* deployed on kubernetes.  I don't think that can of worms has been openned.  This is another layer of redundancy which is good, but I think this would require manual scaling.  Kinda on the fence with this one.  Also it allows for handling some of C* disk needs.

**Disk Needs**

One of the critical performance issues with Cassandra is the disk, both throughput and IOPS. Two mount points are recommended, with SSD.  It can run on one mount point for data, but it does like two.  That is handled via the c* yaml file, and would need to be persistent mount points.  Another point about disks is that C* allows for JBOD, and I have no idea if that is applicable with kubernetes.  For dev and test environments, where latency is not a concern this does not apply.  

With the current example is the data being stored within the docker or on the node disk?

**C* HA Needs - Replication Factor**

I would be apprehensive in running C* without DaemonSet.  As you know the replication factor is one of the base configurations that allows for a HA system within Cassandra.  If you get two C* nodes on the same Kubernetes node, you can lose your data redundancy if the kubernetes node fails, and you loose data.  I am also reading about Node Affinity in 1.2, not sure if that is applicable.  You can get around this by having your replication factor high, but really a replication factor of 3 is recommended.  Using persistent disk may allow for this, but not sure.

**Versions**

Which version of C* is gcr.io/google-samples/cassandra:v8 running? What version of java? openJDK 8 is needed, and I don't think Oracle JDK is on the table.

**Upgrades**

Not sure how C* would handle a `gcloud container clusters upgrade`.  Need more feedback from you Kubernetes gurus :)  Also how would C* handle a rolling update.  Depending on the version jump, you may need to stop the entire cluster.

**Multiple DCs**

Any thoughts about running multiple data centers?  Probably not yet is the answer.

**Synopsis**

Sorry to be long winded, so let me wrap up :)

- documentation on recovery process
- persistent disk, how can this be done?
- multiple disks because C* likes it - it is all about the Disk
- run one C* node on one kubernetes node - you can lose your HA
- version of C* that is in the demo
- kubernetes upgrade
- multiple DCs

**Comment by chrislovecnm** [comment]
May have answered one of my own questions.  The seed provider is compiled against C* 2.0.11, but the docker is being built against the latest debian 21x repo ... we need one of those updated :)

**Comment by chrislovecnm** [comment]
In regards to documenting the seed provider.  Multiple environment vars, and a service account are used.

This is the mentioned here: https://github.com/kubernetes/kubernetes/issues/17724

**Comment by amygdala** [comment]
Just linking in: https://github.com/kubernetes/kubernetes/issues/24030 also.

The PR that removes the pod adoption should be in shortly- https://github.com/kubernetes/kubernetes/pull/22810

**Comment by chrislovecnm** [comment]
And here is another ...https://github.com/kubernetes/kubernetes/issues/24286

@amygdala / @bgrant0607 can we be special and get a Cassandra Label - will make my life a BUNCH easier!!

**Comment by chrislovecnm** [comment]
So we do have a base problem with the way the custom seed provider works.  See https://github.com/kubernetes/kubernetes/issues/24290

It is making all of the Cassandra Nodes a seed.  Ugh.

**Comment by chrislovecnm** [comment]
@bgrant0607 and @amygdala we are running a really old version of C* ... Couple of ideas, how about using the https://hub.docker.com/_/cassandra/ docker.  Also the SeedProvider needs to be built against the correct version of source :smile:

**Comment by bgrant0607** [comment]
Thanks, @chrislovecnm. I created an area/example/cassandra label and applied it to issues I know about, including the one you just filed.

**Comment by chrislovecnm** [comment]
@bgrant0607 any opinion on running C* official docker??  I can file an issue.  Also I have a couple of other tweaks for the build as a stop gap.  We need to set the cassandra jar version.  It is building against an old version.

**Comment by chrislovecnm** [comment]
@bgrant0607 https://github.com/kubernetes/kubernetes/pull/24296  Oh please comment away :)

**Comment by chrislovecnm** [comment]
We got another one :)

https://github.com/kubernetes/kubernetes/issues/24493

@amygdala @bgrant0607 I am going to start to work as soon as I get the rest of the example working, but we need to upgrade

**Comment by chrislovecnm** [comment]
@amygdala @bgrant0607 so how about some good news 😄 

All the examples are working on the master branch, as long as you create the service.  We need to say why you need the service, in regards to the SeedProvider.  I also ran a quick smoke test:

`kubectl --kubeconfig="./kubeconfig" describe svc cassandra` to get the cql port
`kubectl --kubeconfig="./kubeconfig" exec -ti cassandra-szdkn -- bash` start a shell
`cqlsh 10.2.34.2` one of the nodes that I got from `describe svc`

Ran the following inside of `cqlsh`

```
CREATE KEYSPACE demo WITH replication = {
  'class': 'SimpleStrategy',
  'replication_factor': '3'
};

use demo;

CREATE TABLE users (
firstname text, 
lastname text, 
age int, 
email text, 
city text, 
PRIMARY KEY (lastname));

INSERT INTO users (firstname, lastname, age, email, city) VALUES ('John', 'Smith', 46, 'johnsmith@email.com', 'Sacramento'); 
INSERT INTO users (firstname, lastname, age, email, city) VALUES ('Jane', 'Doe', 36, 'janedoe@email.com', 'Beverly Hills'); 

SELECT * FROM users;
```

Might be good thing to include in the demo.  As well as how does k8s external access.

Using DeamonSet is good and bad.  All of the nodes start at the same time, and scaling down would be fun.  BTW the settings in the yaml files yah C* crawling.  2 cpu and 1024M is much friendlier.

**Comment by chrislovecnm** [comment]
@bgrant0607 just put https://github.com/kubernetes/kubernetes/pull/24885 in ... Fixes a bunch of the open issues that I have.  Good start on a stable `SeedProvider` for PetSet 😄

**Comment by bgrant0607** [comment]
Thanks for all your work on this, @chrislovecnm

**Comment by chrislovecnm** [comment]
@bgrant0607 can we close this ... or at least rename this??  Cause it aint really broken 😄

**Comment by bgrant0607** [comment]
@chrislovecnm 

The e2e test is still continuously failing:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/examples.go:257
Expected error:
    <*errors.errorString | 0xc8203deb40>: {
        s: "Failed to get entpoints for e2e-tests-examples-99qtt/cassandra",
    }
    Failed to get entpoints for e2e-tests-examples-99qtt/cassandra
not to have occurred
```

Are #17275 and #10931 still problems?

Are there individual issues filed for outstanding AIs mentioned above?

**Comment by chrislovecnm** [comment]
@bgrant0607 I am having issues getting a v9 image in that works when it is deployed to gcr.io - I have had two different PRs go in that both worked locally, and then don't work when they built by someone else. So currently the v9 image is not stable. I have not had time to look at the e2e tests and they need to be improved as well. I have put in https://github.com/kubernetes/kubernetes/issues/25889#issuecomment-221163864 to design how to improve this process.

The two issues that you mentioned, should probably be closed.

**Comment by chrislovecnm** [comment]
@bgrant0607 the PR is ready for merge on the v9 image. What do we need to do to close this out? I have another issues open for e2e tests which impacts all examples.

**Comment by bgrant0607** [comment]
@chrislovecnm Has the v9 image been pushed? If not, @jeffmendoza should be able to help with that.

cc @kubernetes/examples

**Comment by chrislovecnm** [comment]
@bgrant0607 yep it is up ...

**Comment by bgrant0607** [comment]
Great. Closing in favor of more specific issues, then.

**Comment by kenng** [comment]
Since PetSet is still under alpha, so we got to wait for it to be out of alpha. 

While using daemonSet, it seem that read only volume is allowed. So I am thinking to create a C* container which use persistent disk in the one of the node in order to backup the data, besides running the daemonset for C*. Is it good to do so? Not sure if this is the right place to ask...
